### PR TITLE
[RISCV][GISEL] Legalize G_INSERT_SUBVECTOR and G_EXTRACT_SUBVECTOR for scalable vectors

### PR DESCRIPTION
**Pull request by michaelmaitland**
This PR legalizes G_INSERT_SUBVECTOR and G_EXTRACT_SUBVECTOR for scalable vectors. It is heavily based off the SelectionDAG code.

### CODE REVIEW

**Comment by arsenm** [comment]
I still think we should work towards entirely removing G_INSERT and G_EXTRACT

**Comment by michaelmaitland** [comment]
> I still think we should work towards entirely removing G_INSERT and G_EXTRACT

Can you give me a better idea what you had in mind instead?

**Comment by arsenm** [comment]
> > I still think we should work towards entirely removing G_INSERT and G_EXTRACT
> 
> Can you give me a better idea what you had in mind instead?

Always use total G_MERGE_VALUES / G_UNMERGE_VALUES, G_BUILD_VECTOR, etc. Arbitrary bit offset is impossibly general and hard to work with. It's as constrained as a finally selected instruction, so it's not really that useful for legalization

**Comment by tschuett** [comment]
If this becomes more than "I still think ...", can we move this one floor up to discourse for awareness? E.g., we have to update LegalizerHelper to not produce the G_INSERT/G_EXTRACT family anymore. Put a deprecated comment in GenericOpcodes.td. Put a comment in https://llvm.org/docs/GlobalISel/GenericOpcode.html. And more ...

**Comment by michaelmaitland** [comment]
The purpose of this patch is so that it can be used by #108250. I propose that we accept this patch so we can move #108250 through. Then, we should look to revise #108250 to not rely on G_INSERT and G_EXTRACT. I do not have the bandwidth to rewrite #108250 without G_INSERT and G_EXTRACT at the moment.

File: llvm/lib/Target/RISCV/RISCVInstrGISel.td [hunk 1]
```diff
@@ -41,6 +41,82 @@ def G_VMCLR_VL : RISCVGenericInstruction {
 }
 def : GINodeEquiv<G_VMCLR_VL, riscv_vmclr_vl>;
 
+// Pseudo equivalent to a RISCVISD::VFMV_S_F_VL
+def G_VFMV_S_F_VL : RISCVGenericInstruction {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrGISel.td:45]: This isn't used?

File: llvm/lib/Target/RISCV/RISCVInstrGISel.td [hunk 2]
```diff
@@ -41,6 +41,82 @@ def G_VMCLR_VL : RISCVGenericInstruction {
...
+  let OutOperandList = (outs type0:$dst);
+  let InOperandList = (ins type0:$vec, type1:$scalar, type2:$vl);
+  let hasSideEffects = false;
+}
+def : GINodeEquiv<G_VFMV_S_F_VL, riscv_vfmv_s_f_vl>;
+
+// Pseudo equivalent to a RISCVISD::VFMV_V_F_VL
+def G_VFMV_V_F_VL : RISCVGenericInstruction {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrGISel.td:53]: This isn't used?

File: llvm/lib/Target/RISCV/RISCVInstrGISel.td [hunk 3]
```diff
@@ -41,6 +41,82 @@ def G_VMCLR_VL : RISCVGenericInstruction {
...
+  let OutOperandList = (outs type0:$dst);
+  let InOperandList = (ins type0:$vec, type1:$scalar, type2:$vl);
+  let hasSideEffects = false;
+}
+def : GINodeEquiv<G_VFMV_V_F_VL, riscv_vfmv_v_f_vl>;
+
+// Pseudo equivalent to a RISCVISD::VMV_S_X_VL
+def G_VMV_S_X_VL : RISCVGenericInstruction {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrGISel.td:61]: This isn't used?

File: llvm/lib/Target/RISCV/RISCVInstrGISel.td [hunk 4]
```diff
@@ -41,6 +41,82 @@ def G_VMCLR_VL : RISCVGenericInstruction {
...
+  let OutOperandList = (outs type0:$dst);
+  let InOperandList = (ins type0:$vec, type1:$scalar, type2:$vl);
+  let hasSideEffects = false;
+}
+def : GINodeEquiv<G_VMV_S_X_VL, riscv_vmv_s_x_vl>;
+
+// Pseudo equivalent to a RISCVISD::VMV_V_X_VL
+def G_VMV_V_X_VL : RISCVGenericInstruction {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrGISel.td:69]: This isn't used?

**Comment by arsenm** [comment]
> If this becomes more than "I still think ...", can we move this one floor up to discourse for awareness? E.g., we have to update LegalizerHelper to not produce the G_INSERT/G_EXTRACT family anymore. Put a deprecated comment in GenericOpcodes.td. Put a comment in https://llvm.org/docs/GlobalISel/GenericOpcode.html. And more ...


I was working on that last time I had time to look at globalisel, probably 3 years ago at this point. For the G_INSERT_VECTOR_ELT/G_EXTRACT_VECTOR_ELT case, it's easy to just directly handle the constant operand case

**Review by arsenm — COMMENTED** [review]
Title is wrong, this is not handling G_INSERT_SUBVECTOR or G_EXTRACT_SUBVECTOR. This only handles G_INSERT / G_EXTRACT

**Comment by michaelmaitland** [comment]
> Title is wrong, this is not handling G_INSERT_SUBVECTOR or G_EXTRACT_SUBVECTOR. This only handles G_INSERT / G_EXTRACT

Title is right, code is wrong. I implemented the logic of G_INSERT_SUBVECTOR and G_EXTRACT_SUBVECTOR for G_INSERT and G_EXTRACT by accident. Working on a fix. I think this should assuage the concern of wanting to remove G_INSERT and G_EXTRACT. My other patch that is stacked on this should be using G_INSERT_SUBVECTOR and G_EXTRACT_SUBVECTOR too.

**Pull request closed without merging**